### PR TITLE
[WIP] No-copy semantics for large memoryviews

### PR DESCRIPTION
**Pull request by ogrisel**
Here is a PR that backports some of the work done for large bytes in upstream CPython 3.7 or 3.8 in the following PR: https://github.com/python/cpython/pull/4353.

The goal is to make the `memoryview` support of cloudpickle benefit from it and implement nocopy semantics and later any nested numpy arrays datastructures (e.g. pandas dataframes, scipy sparse arrays, large scikit-learn RandomForestClassifier...).

In particular, this would make it possible for dask distributed to spill any large numpy-based datastructure to disk without making any temporary in-memory copy on workers that are close to their memory usage limit.

**Please do not merge this PR while the upstream CPython PR is still under review.**

- [x] prevent `_memoryview_from_bytes` to ever mutate single bytes interned  bytes instances,
- [x] add a battery of unittests to check edge cases with bytes mutability,
- [x] take the numpy serializer out of the test suite and make it available to users,
- [ ] fix pickling for numpy arrays with the object dtype,
- [x] wait for the final review of https://github.com/python/cpython/pull/4353,
- [ ] disable memoization of bytes when serializing a non-contiguous memoryview,
- [ ] refactor the new tests to make them less redundant.

/cc @pitrou @mrocklin.

### CODE REVIEW

**Comment by ogrisel** [comment]
@HyukjinKwon I would love to get your feedback on this PR if you have the time to review it. I still have a couple of things this to check:

- preserve the shape and strides of the memoryview themselves (instead of just at the numpy array level).
- I would like to see if we can get nocopy semantics for pandas and scipy sparse matrices for free (early quick tests show that it does not seem to work for pandas, we have to investigate).

Question:

- shall we provide the nocopy memoryview-based reducer for numpy arrays by default (instead of just providing an example?

**Comment by pitrou** [comment]
@ogrisel, I would hold on on this PR before the CPython changes get accepted and merged. In particular, one should study the possible issues with memoryview lifetime.

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -249,6 +252,93 @@ def _walk_global_ops(code):
                 yield op, instr.arg
 
 
+def _memoryview_from_bytes(data, format, readonly):
+    if not readonly:
```
Comment (llllllllll) [inline cloudpickle/cloudpickle.py]: I have written something very similar for pandas: https://github.com/pandas-dev/pandas/blob/master/pandas/util/move.c#L160

The test suite exercises some interesting edge cases we may want to replicate: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/util/test_util.py#L324

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -249,6 +252,93 @@ def _walk_global_ops(code):
...
+        # Forcibly render the recently unpickled bytes buffer writable:
+        # this is a violation of the immutability of a Python bytes object but
+        # this should be safe because:
+        # - data is a temporary variable that has just been allocated from
+        #   reading from the pickle stream and will never be used outside of
+        #   the scope of this reducer.
+        # - bytes objects are not subject to interning.
```
[the remark this answers is not in the document]
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Hum. I concluded this from the following:

```
>>> sys.intern(b'x')
Traceback (most recent call last):
  File "<ipython-input-5-02fdb271b026>", line 1, in <module>
    sys.intern(b'x')
TypeError: intern() argument 1 must be str, not bytes
```

Do you know the maximum length of interned bytes objects?

File: cloudpickle/cloudpickle.py (hunk 1 above)
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Thanks I will have a look.

**Comment by HyukjinKwon** [comment]
Thanks @ogrisel. I have been trying to follow it. Will try to take a look to check / double check bit by bit when I have some time as well, probably from the next week.

File: cloudpickle/cloudpickle.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Apparently (from experiments), only instances with zero or one-byte length are "interned" although I could not google that this is supposed to be officially the case. I am having a hard time understanding the C code of the `bytesobject.c` file.

[the remark this answers is not in the document]
Comment (llllllllll) [reply cloudpickle/cloudpickle.py]: In pandas we just check the refcount of the object. Regardless of how interning happens, this will be safe if the refcount is exactly 1 AND you steal the reference. We just take a copy if the refcount is not 1. If the refcount is > 1 and the len is > 1, we raise a performance warning stating that something odd happened and a copy was done; however that doesn't happen. We don't warn under len 1 because a copy of 1 or two bytes is trivial.

[the remark this answers is not in the document]
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: In cloudpickle we have to stay 100% pure python (hence this hack in ctypes). Furthermore we have to respect the pickle protocol and this function is called by the load function when hitting the REDUCE opcode. At this point the `bytes` buffer has already been read and is poped from the unpickler stack, but it's still in the lexical scope of the caller function (`load_reduce` IIRC).

However this bytes object will never be used anywhere else, so I am still quite confident that this ctypes hack, when called in the context of the unpickler is "safe".

[the remark this answers is not in the document]
Comment (pitrou) [reply cloudpickle/cloudpickle.py]: @ogrisel the language doesn't guarantee anything here, so you are on your own if you try to mutate a bytes object without taking the kind of precaution @llllllllll talks about.

[the remark this answers is not in the document]
Comment (llllllllll) [reply cloudpickle/cloudpickle.py]: I am not saying that we need to use the C implementation shown in pandas, I am saying that we should copy the refcount checks that it does. We may also want to try harder to hide the mutable bytes object

[the remark this answers is not in the document]
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Thinking more about it, it might be possible to get the refcount down from 2 to 1 by nesting 2 reducers and doing the "mutate_move" in the wrapping reducer. I can also try to hide the `_base` reference with a readonly property.

[the remark this answers is not in the document]
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: I have pushed a fix. It still lacks a more complete test harness though. Will work on that next week.

File: cloudpickle/cloudpickle.py [hunk 3]
```diff
@@ -249,6 +252,138 @@ def _walk_global_ops(code):
...
+    the bytes object. If readonly is False, a new mutable buffer is always
+    allocated.
+    """
+    # If the following call to sys.getrefcount returns 2 it means that there is
+    # one reference from data_holder and one from the temporary arguments
+    # datastructure of the sys.getrefcount call. It is therefore safe to reuse
+    # the memory buffer of the bytes object as writeable buffer to back the
+    # memoryview: this buffer is no longer unreachable from anywhere else.
+    if hasattr(sys, 'getrefcount'):
+        safe_to_mutate = sys.getrefcount(data_holder[0]) <= 2
```
Comment (llllllllll) [inline cloudpickle/cloudpickle.py]: When making calls like this, it may be nice to call out what the references are:
1. The reference owned by `data_holder`
2. The reference pushed onto the data stack from the expression `data_holder[0]`

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: This is what I explained in the above comment. Also, I reorganized the code to make it more readable and easier to test.

**Comment by ogrisel** [comment]
@llllllllll thanks for the feedback. I think I have addressed the safety concerns in the latest batch of commits. This ctypes-based implementation should mirror all the checks done in C in pandas.

**Comment by llllllllll** [comment]
This is really impressive work!

**Comment by ogrisel** [comment]
This PR was a fun learning experience but it's definitely a hack. I think it's better to wait for PEP 574 to be implemented in upstream Python and adopted by numpy (see: https://github.com/numpy/numpy/issues/11161).
